nemo/collections/asr/parts/utils/transcribe_utils.py

```diff
@@ -282,14 +284,26 @@ def prepare_audio_data(cfg: DictConfig) -> Tuple[List[str], bool]:
             logging.error(f"The input dataset_manifest {cfg.dataset_manifest} is empty. Exiting!")
             return None
 
-        all_entries_have_offset_and_duration = True
-        for item in read_and_maybe_sort_manifest(cfg.dataset_manifest, try_sort=cfg.presort_manifest):
-            if not ("offset" in item and "duration" in item):
-                all_entries_have_offset_and_duration = False
-            audio_key = cfg.get('audio_key', 'audio_filepath')
-            audio_file = get_full_path(audio_file=item[audio_key], manifest_file=cfg.dataset_manifest)
-            filepaths.append(audio_file)
+        audio_key = cfg.get('audio_key', 'audio_filepath')
+
+        with NamedTemporaryFile("w", suffix=".json", delete=False) as durations_f, open(cfg.dataset_manifest, "rt") as f:
+            for item in f:
+                item = json.loads(item)
+                item["audio_filepath"] = get_full_path(item["audio_filepath"], cfg.dataset_manifest)
+                if item.get("duration") is None:
+                    sound = sf.SoundFile(item["audio_filepath"])
+                    duration = sound.frames / sound.samplerate
+                    item["duration"] = duration
```

Comment:
soundfile reads full file when duration is not mentioned, is it chosen here for performance?

Reply:
We need to get the duration of the audio file if we want to sort by duration. Sorting by duration can speed you up by a factor of two or so, IIRC. (I have the real data if you want to see it.) I'm not sure if I want to keep the change. It turns out that simply loading these audio files serially just to get their duration is slower than actually transcribing the data, annoyingly...

The duration fields are all "null" for the dataset I used here: https://bc.ngc.nvidia.com/datasets/1618966

So I threw this shim code in.

It is probably better to just update the duration field and make a new version of that dataset.

And throw an exception when someone wants to sort by duration but there is no duration field.

Reply:
yeah reason I asked is, unfortunately reading and writing is where most of the time spent.

Reply:
It's more complicated than that: if the format has a header, such as WAV or FLAC, soundfile will only read the header unless you call `.read()` on that object. But if it's something like MP3, it may read the whole thing.